src/components/FirmwareUpdate/index.tsx

```diff
@@ -0,0 +1,240 @@
+          );
+          return { step: "confirmUpdate" };
+        case "flashingMcu":
+          return {
+            step: "flashingMcu",
+            progress: event.progress,
+            installing: event.installing,
+          };
+        case "firmwareUpdated":
+          return { step: "firmwareUpdated" };
+        case "error":
+          return { step: "error", error: event.error };
+        case "reset":
+          return {
+            step: event.wired ? "confirmRecoveryBackup" : "error",
+            progress: undefined,
+            error: event.wired ? undefined : BluetoothNotSupportedError,
+            installing: undefined,
+          };
+        default:
+          return { ...state };
+      }
+    },
+    [t],
+  );
+
+  const [state, dispatchEvent] = useReducer(fwUpdateStateReducer, {
+    step: device.wired ? "confirmRecoveryBackup" : "error",
+    progress: undefined,
+    error: device.wired ? undefined : BluetoothNotSupportedError,
+    installing: undefined,
+  });
+
+  const { step, progress, error, installing } = state;
+
+  const onReset = useCallback(() => {
+    dispatchEvent({ type: "reset", wired: device.wired });
+    dispatch(clearBackgroundEvents());
+    NativeModules.BackgroundRunner.stop();
+  }, [dispatch]);
+
+  // only allow closing of the modal when the update is not in an intermediate step
+  const canClose =
+    step === "confirmRecoveryBackup" ||
+    step === "firmwareUpdated" ||
+    step === "error";
+
+  const onTryClose = useCallback(
+    (restoreApps: boolean) => {
+      if (canClose) {
+        onClose(restoreApps);
+      }
+    },
+    [canClose],
+  );
+
+  const onCloseAndReinstall = useCallback(() => onTryClose(true), [onTryClose]);
+  const onCloseSilently = useCallback(() => onTryClose(false), [onTryClose]);
+
+  useEffect(() => {
+    // reset the state whenever we re-open the modal
+    if (isOpen) {
+      onReset();
+    }
+  }, [isOpen, onReset]);
+
+  useEffect(() => {
+    if (!nextBackgroundEvent) return;
+
+    dispatchEvent(nextBackgroundEvent);
+    dispatch(dequeueBackgroundEvent());
+  }, [nextBackgroundEvent, dispatch, dispatchEvent]);
+
+  useEffect(() => {
+    if (step === "error") {
+      track("FirmwareUpdateError", error ?? null);
+    }
+  }, [step]);
+
+  const launchUpdate = useCallback(() => {
+    if (latestFirmware) {
+      NativeModules.BackgroundRunner.start(
+        device.deviceId,
+        JSON.stringify(latestFirmware),
+        t("FirmwareUpdate.Notifications.preparingUpdate"),
+      );
+      dispatchEvent({ type: "downloadingUpdate", progress: 0 });
+    }
+  }, [latestFirmware]);
+
+  const firmwareVersion = latestFirmware?.final?.name ?? "";
+
+  return (
+    <BottomModal
+      id="DeviceActionModal"
+      noCloseButton={!canClose}
+      isOpened={isOpen}
+      onClose={onCloseSilently}
+      onModalHide={onCloseSilently}
+    >
+      {step === "confirmRecoveryBackup" && (
```

Comment:
Personal preference here but the fact that you are doing `condition && render` in sequence means we are evaluating all the conditions even though they seem to be self excluding, this could be a ternary chain instead which would short the evaluation on the first match

Reply:
I personally feel that the ternary chain is a bit harder to understand in the code. I could also replace it by an IIFE with a switch / case statement. But personally in this case I still favor the `condition && render` for the readability since the cost for comparing strings is very low in JS.